lib/pbench/server/cache_manager.py

```diff
@@ -430,29 +429,51 @@ def get_info(self, path: Path) -> JSONOBJECT:
         return fd_info
 
     @staticmethod
-    def extract(tarball_path: Path, path: str) -> Optional[str]:
-        """Extract a file from the tarball and return it as a string
+    def extract(tarball_path: Path, path: Path) -> Optional[IO[bytes]]:
+        """Returns the file stream which yields the contents of
+        a file at the specified path in the Tarball
+
+        Args:
+            tarball_path: absolute path of the tarball
+            path: relative path within the tarball
+
+        Raise:
+            BadDirpath on failure extracting the file from tarball
+        """
+        try:
+            return tarfile.open(tarball_path, "r:*").extractfile(path)
+        except Exception as exc:
+            raise BadDirpath(
+                f"A problem occurred processing {str(path)!r} from {str(tarball_path)!r}: {exc}"
+            )
 
-        Report failures by raising exceptions.
+    def filestream(self, path: str) -> Optional[dict]:
+        """Returns a dictionary containing information about the target
+        file and file stream
 
         Args:
             path: relative path within the tarball of a file
 
         Raises:
-            MetadataError on failure opening the tarball
             TarballUnpackError on failure to extract the named path
 
         Returns:
-            The named file as a string
+            Dictionary with file info and file stream
         """
+        file_path = Path(self.name) / path
+        info = self.get_info(file_path)
+        info["stream"] = None
+
         try:
-            tar = tarfile.open(tarball_path, "r:*")
-        except Exception as exc:
-            raise MetadataError(tarball_path, exc) from exc
-        try:
-            return tar.extractfile(path).read().decode()
+            if not path:
+                info["stream"] = self.tarball_path.open("rb")
+            if info["type"] == CacheType.FILE:
+                info["stream"] = Tarball.extract(self.tarball_path, file_path)
         except Exception as exc:
-            raise TarballUnpackError(tarball_path, f"Unable to extract {path}") from exc
+            raise TarballUnpackError(
+                self.tarball_path, f"Unable to extract {str(file_path)!r}"
+            ) from exc
+        return info
```

Comment:
If `path` is `None` at line 468, we'll never know because we will have hit a `TypeError` exception at line 463.  (We should have a test for this case!)

The check for the tarball special case needs to be pretty much the first thing in the function.
```python
        if not path:
            info = {
                "type": CacheType.FILE,
                "stream": self.tarball_path.open("rb"),
                # Add whatever other keys we think we need, here.
            }
        else:
            file_path = Path(self.name) / path
            info = self.get_info(file_path)
            if info["type"] == CacheType.FILE:
                try:
                    info["stream"] = Tarball.extract(self.tarball_path, file_path)
                except Exception as exc:
                    raise TarballUnpackError(
                        self.tarball_path, f"Unable to extract {str(file_path)!r}"
                    ) from exc
            else:
                info["stream"] = None

        return info
```
We shouldn't need the `try` block around the `tarball_path.open()` call...right?